Tracks: Add site type details to media events

SiobhyB opened the pull request:
## Description

As a follow on from https://github.com/wordpress-mobile/WordPress-iOS/pull/22328, this PR adds site type details to media tracking events. The addition of these details will help us to further analyse these events in the future. 

More specifically, `is_jetpack`, `is_atomic`, and `is_hosted_at_WPcom` props have been added to the events, so that we can determine the type of site that the media events belonged to.

These new props have been added to the following events:

- `wpios_media_service_upload_started` **||** `jpios_media_service_upload_started`
- `wpios_media_service_upload_successful` **||** `jpios_media_service_upload_successful`
- `wpios_media_service_upload_canceled` **||** `jpios_media_service_upload_canceled`
- `wpios_media_service_upload_failed` **||** `jpios_media_service_upload_failed`


### Testing

* Install the build available in this PR's comments or checkout this branch locally.
* Navigate to the post editor and upload an image from your device's library.
* Before the image finished uploading, tap on it to cancel the process.
* Next, go through the steps to upload another image.
* Before the process finishes, turn off the WiFi so that the upload fails.
* Navigate to the **Live View** tab in Tracks.
* Enter the username associated with your testing account into the **Username, ID, or Email** field. 
* Wait until the events listed in this PR's **description** section show up in your history. _Note, this can take up to 20 minutes._
* When the events show up, verify that the new props are visible when expanding their details. The following table shows the value that the props should display depending on the site type you're using to test.

| Property  | Simple .com | Atomic | Jetpack | Self-hosted (without Jetpack) |
| ------------- | ------------- | ------------- | ------------- | ------------- |
| `is_atomic` | 0 | 1 | 0 | 0 |
| `is_hosted_at_wpcom ` | 1 | 0 | 0 | 0 |
| `is_jetpack` | 0 | 1 | 1 | 0 |

## Regression Notes
1. Potential unintended areas of impact

Any other changes to the way we collect data, beyond adding blog type details to the `*_media_service_upload_*` events, would be unintentional.

2. What I did to test those areas of impact (or what existing automated tests I relied on)

I considered whether it may be valuable to add site type details to all events it was possible for, but decided against this in order to avoid any unintended side effects, especially as I could not find a demand for this type of extra data for other events. By intentionally limited the changes to the `*_media_service_upload_*` events., I hope to avoid any unintended negative impact.

Aside from this, I compared the props for the `*_media_service_upload_*` events before this change and then after to verify there was nothing unexpected.

3. What automated tests I added (or what prevented me from doing so)

There are no existing tests in place for events, so I did not add any. As this PR only attaches a new piece of data to existing events, I did not feel a need to further investigate ways to automate testing for events in general.

<hr />

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

SiobhyB committed:
chore: Add constants for better site type tracking

We want to be able to track various site types with media events.

A `WPAppAnalyticsKeyIsJetpack` constant already exists, which stores an `is_jetpack` string. With this commit, we add to the list of constants to store `is_atomic` and `is_hosted_at_wpcom` strings.

This sets us up to reference those strings elsewhere in the codebase.

SiobhyB committed:
chore: Set up method to generate site type details

The `blogPropertiesToTrack` method generates a dictionary of properties related to the user's blog. Included properties will tell us more information about the type of site the media is being uploaded to i.e. Jetpack, Atomic, or WordPress.com.

SiobhyB committed:
feat: Send site type details as media events fire

The `blogPropertiesToTrack` method, set up in the previous commit, to sent site type details when media tracking events are fired. The existing `withProperties` argument has been utilised in each of the tracking methods to easily pass this info.

SiobhyB committed:
refactor: Remove redundant method

The tracking method that fires when a media upload errors out was refactored in the previous commit. As such, the original method for errors (which was originally added in https://github.com/wordpress-mobile/WordPress-iOS/pull/22328) is no longer used anywhere in the codebase and can be removed. It's removed with this commit.

SiobhyB commented:
> Fixes #
> 
> ### Testing
> Events to test:
> 
> * `wpios_media_service_upload_started` **||** `jpios_media_service_upload_started`
> * `wpios_media_service_upload_successful` **||** `jpios_media_service_upload_successful`
> * `wpios_media_service_upload_canceled` **||** `jpios_media_service_upload_canceled`
> * `wpios_media_service_upload_failed` **||** `jpios_media_service_upload_failed`
> 
> Property	Simple .com	Atomic	Jetpack	Self-hosted (without Jetpack)
> `is_atomic`	0	TBD	TBD	TBD
> `is_hosted_at_wpcom `	1	TBD	TBD	TBD
> `is_jetpack`	0	TBD	TBD	TBD
> ## Regression Notes
> 1. Potential unintended areas of impact
> 2. What I did to test those areas of impact (or what existing automated tests I relied on)
> 3. What automated tests I added (or what prevented me from doing so)
> 
> PR submission checklist:
> 
> * [x]  I have completed the Regression Notes.
> * [x]  I have considered adding unit tests for my changes.
> * [x]  I have considered adding accessibility improvements for my changes.
> * [x]  I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

SiobhyB committed:
Revert "refactor: Remove redundant method"

This reverts commit f6e3e6cc9b0e3876a1f1369212f8d9a26dd7278b.

SiobhyB committed:
fix: Ensure errors are tracked correctly

With my previous commits, I mistakenly reverted changes that were necessary to correctly track media upload errors. These changes now been restored, with some additional tweaks in this commit to ensure this works well with the additional site type details.

SiobhyB committed:
Merge branch 'trunk' into task/add-blog-properties-to-media-events

twstokes commented on WordPress/Classes/Services/MediaService.m:
```diff
@@ -565,4 +567,15 @@ - (RemoteMedia *)remoteMediaFromMedia:(Media *)media fieldsToUpdate:(NSArray<NSS
 
+- (NSMutableDictionary *)blogPropertiesToTrack:(Blog *)blog {
+    NSMutableDictionary *properties = [NSMutableDictionary dictionary];
+    BOOL isJetpackConnected = blog.jetpack != nil && blog.jetpack.isConnected;
```
I may be wrong here @SiobhyB but I was thinking Objective-C will return `nil` even if `blog.jetpack` is `nil`, e.g.:

`BOOL isJetpackConnected = blog.jetpack.isConnected != nil;`

But it'd be good to double-check!

SiobhyB replied:
@twstokes, thanks for this note! I got the following error when testing the suggested line: `Comparison between pointer and integer ('BOOL' (aka '_Bool') and 'void *')`. I think because `isConnected` is a BOOL that can't be compared to `nil`.

To get the line more streamlined though, prompted by your comment, I've made an update in https://github.com/wordpress-mobile/WordPress-iOS/pull/22360/commits/7cb3d7f1fe7198ac0bb8dba06902dc1ad8f7cea4 to:

`BOOL isJetpackConnected = blog.jetpack.isConnected ? YES : NO;`

As I understand, if `blog.jetpack` is `nil`, the above will still safely return `NO`. Let me know if that seems right to you!

twstokes replied:
Ah! Thanks for double-checking - I'm rusty on Objective-C. 😄

In that case, would `properties[WPAppAnalyticsKeyIsJetpack] = @(blog.jetpack.isConnected);` work? If `jetpack` is `nil` shouldn't Objective-C evaluate it to falsy / boolean `NO` like it did with your latest change?

SiobhyB replied:
Ah, yes, you're right! Thank you. I wasn't seeing the forest for the trees there, I updated in https://github.com/wordpress-mobile/WordPress-iOS/pull/22360/commits/12850499ab9f93ff01d72c43a7d723b6115d9ea0.

twstokes replied:
Awesome! I didn't see that solution either initially, but I'm glad we arrived to it!

SiobhyB committed:
Merge branch 'trunk' into task/add-blog-properties-to-media-events

SiobhyB committed:
refactor: Streamline logic around Jetpack check

The commit follows the feedback here: https://github.com/wordpress-mobile/WordPress-iOS/pull/22360#discussion_r1448203508

SiobhyB committed:
refactor: Simplify `isJetpackConnected` assignment

See: https://github.com/wordpress-mobile/WordPress-iOS/pull/22360#discussion_r1448932566

SiobhyB commented:
Going ahead to close this PR following internal discussions with both @jhnstn and @twstokes. We've decided that having the blog ID associated with the events (added in https://github.com/wordpress-mobile/WordPress-iOS/pull/22328) will satisfy our need to track more info about these media events. Using the blog ID, we'll be able to create queries to determine site type, if needs be.
